transport/src/main/java/io/netty/channel/AbstractCoalescingBufferQueue.java

```diff
@@ -0,0 +1,267 @@
+        for (;;) {
+            Object entry = bufAndListenerPairs.poll();
+            if (entry == null) {
+                if (previousBuf != null) {
+                    ctx.write(previousBuf, ctx.voidPromise());
+                }
+                break;
+            }
+            try {
+                if (entry instanceof ByteBuf) {
+                    if (previousBuf != null) {
+                        ctx.write(previousBuf, ctx.voidPromise());
+                    }
+                    previousBuf = (ByteBuf) entry;
+                } else if (entry instanceof ChannelPromise) {
+                    ctx.write(previousBuf, (ChannelPromise) entry);
+                    previousBuf = null;
+                } else {
+                    ctx.write(previousBuf).addListener((ChannelFutureListener) entry);
+                    previousBuf = null;
+                }
+            } catch (Throwable t) {
+                if (pending == null) {
+                    pending = t;
+                }
+            }
+        }
+        if (pending != null) {
+            throw new IllegalStateException(pending);
+        }
+    }
+
+    /**
+     * Calculate the result of {@code current + next}.
+     */
+    protected abstract ByteBuf compose(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf next);
+
+    /**
+     * Calculate the first {@link ByteBuf} which will be used in subsequent calls to
+     * {@link #compose(ByteBufAllocator, ByteBuf, ByteBuf)}.
+     */
+    protected ByteBuf composeFirst(ByteBufAllocator allocator, ByteBuf first) {
+        return first;
+    }
+
+    /**
+     * The value to return when {@link #remove(ByteBufAllocator, int, ChannelPromise)} is called but the queue is empty.
+     * @return the {@link ByteBuf} which represents an empty queue.
+     */
+    protected abstract ByteBuf removeEmptyValue();
+
+    /**
+     * Get the number of elements in this queue added via one of the {@link #add(ByteBuf)} methods.
+     * @return the number of elements in this queue.
+     */
+    protected final int size() {
+        return bufAndListenerPairs.size();
+    }
+
+    private void releaseAndCompleteAll(ChannelFuture future) {
+        readableBytes = 0;
+        Throwable pending = null;
+        for (;;) {
+            Object entry = bufAndListenerPairs.poll();
+            if (entry == null) {
+                break;
+            }
+            try {
+                if (entry instanceof ByteBuf) {
+                    ReferenceCountUtil.safeRelease(entry);
+                } else {
+                    ((ChannelFutureListener) entry).operationComplete(future);
+                }
+            } catch (Throwable t) {
+                pending = t;
+            }
+        }
+        if (pending != null) {
+            throw new IllegalStateException(pending);
+        }
+    }
+
+    private void checkByteBufPreconditions(ByteBuf buf) {
+        if (readableBytes > Integer.MAX_VALUE - buf.readableBytes()) {
```

Comment:
Interesting overflows check.. 👍 
If it were necessary to check also that the `buf.readableBytes()` is not less than zero, it could be writen as:
```java
if (readableBytes + buf.readableBytes() < readableBytes) {
    throw ...
}
```